src/DataTable/tableReducer.js

```diff
@@ -41,21 +41,20 @@ export function tableReducer(state, action) {
         ...state,
         selectedCount: selectedRows.length,
         allSelected: selectedRows.length === rows.length,
-        selectedRows,
+        selectedRows: [...state.selectedRows, ...selectedRows],
```

Comment:
I missed this one. What was the reason for merging these?
 Versus just replacing?

Reply:
Previously this `useEffect` replaced the selections with the subset of `data` that is filtered out by `selectableRowSelected`. `data` is only the current page when using `paginationServer`. So this change is made to avoid wiping out selections from other pages.

https://github.com/MaxInertia/react-data-table-component/blob/0714517d832bc805d0120d6a29c2d7f8f6fb9d98/src/DataTable/DataTable.js#L162-L170

I'm glad you brought this up though, there's a big problem with this: There are situations where we'll end up with duplicate rows in `selectedRows` with a naive merging. Only those that aren't already in `state.selectedRows` need to be merged.

Reply:
@jbetancur What do you think about something like this?

```
[...state.selectedRows, ...selectedRows.filter(row => !state.selectedRows.includes(row))]
```

My only concern is how this will fare when `state.selectedRows` and `selectedRows` are large...

Reply:
The code above the worst case in O(n+m) - which is not horrible but a case that I would prefer to avoid. I know that even if someone has say ~1000 rows I am pretty sure the DOM will run into performance issues before this code (in which case you need to page), but still... it feels weird.

Aside from a for loop which would be negligible in terms of performance, there is Set.

```js
 Array.from(new Set(array))
```
The big gotcha with Set is it will need to be polyfilled for IE11 users so will need to figure that out.

https://stackoverflow.com/questions/39007637/javascript-set-vs-array-performance


Another solution could be to only perform the expensive operation when persistSelectedOnX where we can merge arrays but in any case we default to replace selectedRows to preserve the current performance.


Sorry for the long comment but this is making me think what is this use case looks like in your implementation? Typically UX wise it is recommended to only allow selections on visible items where this is more edge case of server-side pagination to make up for the limits of server-side pagination

Reply:
>Typically UX wise it is recommended to only allow selections on visible items where this is more edge case of server-side pagination to make up for the limits of server-side pagination

This isn't a necessary limitation with server-side pagination in general, it's a limitation with server-side pagination on this table. This change brings the behavior of the table when using server-side pagination in-line with it's behavior when not using server-side pagination.

Currently to get selection working on multiple pages we're forced to query for all our data upfront, which is wasteful.

>Another solution could be to only perform the expensive operation when persistSelectedOnX where we can merge arrays but in any case we default to replace selectedRows to preserve the current performance.

This might be the best option right now. Although, my code in the last comment is actually `O(nm)` since each of the previous selections results in an iteration over the new selections.

Reply:
Yea O(nm) is correct, my error.. O(n+m) might have been better if it were possible.

You have a valid point on the limitation. Another option could be the cache the rows as off page rows and pass the merging off, but perhaps that adds more complexity at the cost of pre-optimization. If someone is paging why would they page 10000 rows (I’m sure someone will log that issue now :))

Anyway I just want to weigh all the options here.

Let’s do this. Just add the code to only to the expensive calculation when this option is specified and add the test as needed to check that the rows are cached on server page and let’s get this merged in.

Sidebar: It would be cool to have Some sample code for this for a working storybook example. Totally not needed for this pr though.